### PR TITLE
Basic implementation of `:properties`

### PR DESCRIPTION
**Pull request by seanl-adg**

### CODE REVIEW

File: lib/style-observer.js [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    };
+
+    /**/
+
+    var pendingStyles = new utils.Set();
+    var sheetToFilterSelectorMap = new utils.WeakMap();
+
+    var anyStyleWasUpdated;
+    var examineStyles = function() {
+        rAFid = undefined;
```
Comment (ameshkov) [inline lib/style-observer.js]: rAF? What's rAF?

File: lib/style-observer.js [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+                return element.matches(selectors.join(','));
+            }
+        });
+    };
+
+    return {
+        addIgnoredSheets: addIgnoredSheets,
+        initialize: initialize,
+        extendSizzle: extendSizzle,
+        getSelector: getSelector
```
Comment (ameshkov) [inline lib/style-observer.js]: Why is `getSelector` exposed? I don't see where it's used externally.

File: lib/style-observer.js [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    var reProperty = /[A-Za-z0-9\-]+/;
+    var registerStylePropertyFilter = function (filter) {
+        filter = filter.trim();
+        var property, re;
+        if (reRegexRule.test(filter)) {
+            filter = filter.slice(1, -1);
+            var index = filter.indexOf(':');
+            property = filter.slice(0, index).trim();
+            if (!reProperty.test(property)) {
+                console.warn('Unsupported regex usage of :properties(' + filter + ')');
```
Comment (ameshkov) [inline lib/style-observer.js]: Let's extract it to `utils.logger` at least

File: lib/extended-css.js [hunk 4]
```diff
@@ -152,100 +195,43 @@ var ExtendedCss = function (styleSheet) { // jshint ignore:line
...
+        observeDocument(mainCallback);
     };
 
-    /**
-     * Applies extended CSS rules
-     */
     var apply = function () {
         applyRules(rules);
         observe();
+        StyleObserver.initialize();
```
Comment (ameshkov) [inline lib/extended-css.js]: It shouldn't be initialized if there're no `:properties` rules

**Comment by ameshkov** [comment]
Your code is usually well-documented, but not this time.

**Comment by seanl-adg** [comment]
I wanted a quick prototype, I will add detailed documentations later. It is possible that we will need to change some functions and then a block of comment will become unnecessary.

**Comment by ameshkov** [comment]
Oh, and one more thing. As I understand now how it works, I disagree with this point:

> but I think the matchability of `:properties` rule should not depend on an order of which properties are mentioned in the stylesheet

We do this for compatibility mostly so it'd be better to use the same exact algorithm as they do.

**Comment by ameshkov** [comment]
> I wanted a quick prototype

Well, it's not bad for a prototype :)

**Comment by seanl-adg** [comment]
Please do not post things that are already discussed in the chats as if it is being told for the first time :)

File: lib/style-observer.js [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+            }
+            Array.prototype.push.apply(selectors, map[filter]);
+        }
+        getSelectorCache[filter] = selectors;
+        return selectors;
+    };
+
+    var parsedFilters = [];
+
+    window.parsedFilters = parsedFilters;
```
[the remark this answers is not in the document]
Comment (seanl-adg) [reply lib/style-observer.js]: For debugging purposes

File: lib/style-observer.js (hunk 2 above)
Comment (seanl-adg) [reply lib/style-observer.js]: For debugging purposes

File: lib/utils.js [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        }
+
+        return regex;        
+    };
+
+    return createRegexText;
+})();
+
+// https://github.com/AdguardTeam/FingerprintingBlocker/blob/master/src/shared/url.ts#L64
+utils.createLocation = function (href) {
```
[the remark this answers is not in the document]
Comment (seanl-adg) [reply lib/utils.js]: 2. It will increase the code size.

File: lib/utils.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            this.wrappedCallback();
+        }
+    };
+
+    return AsyncWrapper;
+})();
+
+utils.defineProperty = Object.defineProperty;
+
+utils.WeakMap = typeof WeakMap !== 'undefined' ? WeakMap : (function() {
```
[the remark this answers is not in the document]
Comment (seanl-adg) [reply lib/utils.js]: `Set` is mine.

File: lib/utils.js [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    utils.defineProperty(Set.prototype, 'size', {
+        get: function() {
+            return this.keys.length;
+        }
+    });
+
+    return Set;
+})();
+
+utils.matchesPropertyName = (function() {
```
[the remark this answers is not in the document]
Comment (seanl-adg) [reply lib/utils.js]: I think its purpose is generic enough, and can be used via other 'modules'.

**Comment by seanl-adg** [comment]
@ameshkov Unit tests pass on all browser (Edge, Chrome, Firefox, IE9~) but phantomjs. Except that, I think the pull request is ready to be reviewed. Debugging problems with phantomjs is extremely hard, can we skip it? Also, since phantomjs is close to Safari, can you do a bit of debugging on your end?

**Comment by seanl-adg** [comment]
There are a few more things to consider before actually merging this:

 - A presense of `:properties` rule to be applied to a page will require it to lookup the whole stylesheets in the page on the page load time. I need to examine its cost.

 - Easylist contains several rules like
```
gelbooru.com##[-abp-properties='*: *;base64,']
```
 - In this implementation, such rules will be very inefficient. It is not that particularly inefficient in ABP, in comparison with other `-abp-properties` rules, because of their algorithm for applying style. However, we are pretty much restricted with the matching algorithm of Sizzle. We'd better think about how to optimize such cases.

**Comment by seanl-adg** [comment]
> for instance waitFor(conditionFunc, timeout, callbackFunc)
You need a comment for your function.

**Comment by ameshkov** [comment]
> You need a comment for your function.

There's no such function (yet) :)

A function that waits for the `conditionFunc` to return true or for the specified timeout to exceed. Once any of that happens, it calls the `callbackFunc` and passes the result (either true or false when the timeout was exceeded).

**Comment by seanl-adg** [comment]
You just threw out a something that looks like a function declaration as a feedback, how could I know what this glorious function is supposed to without further explanation?

File: Gruntfile.js [hunk 9]
```diff
@@ -26,12 +28,13 @@ module.exports = function(grunt) {
       },
       dist: {
         src: [
-          'lib/**.js',
+          'dist/utils.js',
+          'dist/**.js',
```
Comment (ameshkov) [inline Gruntfile.js]: Won't we end up with duplicated utils.js in the concatenation result?

File: Gruntfile.js [hunk 10]
```diff
@@ -49,8 +52,9 @@ module.exports = function(grunt) {
       options: {
         curly: true,
         eqeqeq: true,
+        esversion: 6,
         immed: true,
-        latedef: true,
+        // latedef: true,
```
Comment (ameshkov) [inline Gruntfile.js]: Do not comment out, remove it

File: lib/extended-css.js [hunk 11]
```diff
@@ -27,30 +27,80 @@ var ExtendedCss = function (styleSheet) { // jshint ignore:line
...
+     * @returns Mutation observer used to protect attribute or null if there's nothing to protect
+     */
+    var protectStyleAttribute = function (node) {
+        if (!utils.MutationObserver) { return null; }
+        var protectionObserver = new utils.MutationObserver(protectionFunction);
+        protectionObserver.observe(node, protectionObserverOption);
+        return protectionObserver;
     };
 
+    /** Should be clear what it is. */
```
Comment (ameshkov) [inline lib/extended-css.js]: Make sure this is the last time you're making your point via the production code. This is childish, dangerous if tolerated, and unexpected of you.

File: lib/extended-css.js [hunk 12]
```diff
@@ -150,102 +182,50 @@ var ExtendedCss = function (styleSheet) { // jshint ignore:line
      * @param rules Rules to apply
      */
     var applyRules = function (rules) {
+        // var start = performance.now();
```
Comment (ameshkov) [inline lib/extended-css.js]: I guess that's the right place to record timings.

What we need:

1. The total number of times applyRules was called
2. Total time spent
3. Avg time per call
4. Std deviation

On a side note, it sometimes useful to see if the page is slow due to the ExtCss rules. Do you see any simple way to enable timings recording from the extension side?

File: lib/extended-css.js (hunk 11 above)
Comment (seanl-adg) [reply lib/extended-css.js]: You seem to be unable to refrain from posting such a silly comment publicly, here is what I propose.
 - Stop discussing pull requests here in GitHub, and move all reviewing stuff to Bitbucket as we do with the browser extension.

**Comment by seanl-adg** [comment]
We should focus on the code here, not judging and flaming each other. It is really disappointing that we have to do this on every iteration.

**Pull request merged**